frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+// limitations under the License.
+
+//! # Glutton Pallet
+//!
+//! Pallet that consumes `ref_time` and `proof_size` of a block. Based on the
+//! `Compute` and `Storage` parameters the pallet consumes the adequate amount
+//! of weight.
+//!
+//! NOTE: This is only meant to be used for testing.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+use blake2::{Blake2b512, Digest};
+use frame_support::{defensive, pallet_prelude::*, weights::WeightMeter};
+use frame_system::pallet_prelude::*;
+use sp_runtime::{traits::Zero, Perbill, Saturating};
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Weight information for this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		/// The pallet has been initialized by root.
+		PalletInitialized,
+		/// The computation limit has been updated by root.
+		ComputationLimitSet { compute: Perbill },
+		/// The storage limit has been updated by root.
+		StorageLimitSet { storage: Perbill },
+	}
+
+	#[pallet::storage]
+	pub(crate) type Compute<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(crate) type Storage<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(super) type TrashData<T: Config> = StorageMap<_, Blake2_128Concat, u32, u32, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		/// The initial percentage of the `ref_time` to waste.
+		pub compute: Perbill,
+		/// The initial percentage of the `proof_size` to consume.
+		pub storage: Perbill,
+	}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self { compute: Default::default(), storage: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			Compute::<T>::set(self.compute);
+			Storage::<T>::set(self.storage);
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(
+				!T::WeightInfo::waste_ref_time(1).ref_time().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+			assert!(
+				!T::WeightInfo::waste_proof_size_none(1).proof_size().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+		}
+
+		fn on_idle(_: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut meter = WeightMeter::from_limit(remaining_weight);
+			if !meter.check_accrue(T::DbWeight::get().reads(2)) {
```

Comment:
Using here `reads` is wrong in the future AFAIK? Because when we also track the proof size, this call can not account for the used proof size. TLDR: We probably need some benchmark that doesn't waste proof size nor ref time.

Reply:
I don't think I understand this correctly. Why wouldn't we account `T::DbWeight::get().reads(2)` for the used proof size? We have this check because we are reading `Storage` and `Compute` storage values in the next lines.

In case we don't have enough `remaining_weight` to read these two storage values we return `T::WeightInfo::empty_on_idle()` which basically has weight of zero.